### PR TITLE
Spi dma take3

### PR DESCRIPTION
**Pull request by bobmcwhirter**

### CODE REVIEW

File: embassy-stm32/src/spi/v1.rs [hunk 1]
```diff
@@ -128,9 +141,150 @@ impl<'d, T: Instance> Spi<'d, T> {
...
+    where
+        Tx: TxDmaChannel<T>,
+    {
+        unsafe {
+            T::regs().cr1().modify(|w| {
+                w.set_spe(false);
+            });
+            T::regs().cr2().modify(|reg| {
+                reg.set_rxdmaen(true);
+            });
```
Comment (thalesfragoso) [inline embassy-stm32/src/spi/v1.rs:156]: is this needed here ?

File: embassy-stm32/src/spi/v1.rs [hunk 2]
```diff
@@ -128,9 +141,150 @@ impl<'d, T: Instance> Spi<'d, T> {
...
+            T::regs().cr1().modify(|w| {
+                w.set_spe(false);
+            });
+        }
+
+        Ok(())
+    }
+
+    #[allow(unused)]
+    async fn read_write_dma_u8(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Error>
```
Comment (thalesfragoso) [inline embassy-stm32/src/spi/v1.rs:231]: Should we check for the buffers being the same size ? What happens if `read.len() > write.len()` ?

File: embassy-stm32/src/spi/v3.rs [hunk 3]
```diff
@@ -161,9 +175,167 @@ impl<'d, T: Instance> Spi<'d, T> {
...
+            T::regs().cfg1().modify(|reg| {
+                reg.set_rxdmaen(true);
+            });
+        }
+
+        let rx_request = self.rxdma.request();
+        let rx_src = T::regs().rxdr().ptr() as *mut u8;
+        let rx_f = self
+            .rxdma
+            .read(rx_request, rx_src, &mut read[0..write.len()]);
```
Comment (thalesfragoso) [inline embassy-stm32/src/spi/v3.rs:295]: This seems to make more sense if compared to the other two implementations.

**Review by thalesfragoso — COMMENTED** [review]
Thanks again! I left some comments that I think apply to both v1 and v2.

File: embassy-stm32/src/spi/v3.rs (hunk 3 above)
Comment (thalesfragoso) [reply embassy-stm32/src/spi/v3.rs:295]: Maybe we should have some indications that says that if `read.len() > write.len()` then `read` won't be fully filled ? Or maybe just require the lengths to be equal ?

Comment (thalesfragoso) [reply embassy-stm32/src/spi/v3.rs:295]: Actually, I think that might panic if `write` is bigger than `read`.

Comment (bobmcwhirter) [reply embassy-stm32/src/spi/v3.rs:295]: Yah I think definitely an assert but I am not sure how I feel about the slicing really. 

I'll double check which DMA directions get enabled in write-only modes on all of them.

Comment (bobmcwhirter) [reply embassy-stm32/src/spi/v3.rs:295]: If read is longer then it never fills and won't complete. Which I hit a few times myself.

File: embassy-stm32/src/spi/v1.rs (hunk 2 above)
Comment (bobmcwhirter) [reply embassy-stm32/src/spi/v1.rs:228]: Added both an assert that read >= write, slicing it down as needed, and documented.

**Comment by bobmcwhirter** [comment]
Also, where do we stand on the string-of-commits vs squash-it-down continuum?

**Review by Dirbaio — APPROVED** [review]
:rocket:

**Pull request merged**